Issue 44 wider repeat

obregos opened the pull request:
- add widerRepeated function to reshape data from exportRecordsTyped into long format
- add widerRepeated test cases

spgarbet commented:
1. Remove the `cat` calls
2. Add in the documentation that it will only convert those with an idvar (e.g. record_id), "redcap_event_name", and "redcap_repeat_instrument" in their fields. Otherwise it will return the data passed in unchanged. 
3. It needs to preserve the "invalid" attribute of what is passed in. Around 39. 
 
```
if(!is.null(attr(Records, "invalid")))
{
  attr(Records_wide, "invalid") <- attr(Records, "invalid")
  attr(Records_wide, "invalid")$row <- NA # Since we can't tell where these landed
}
```

spgarbet commented:
Oh, snap! I just remembered that there can be attributes on columns in the data frame as well and all those need to come over. `reshape` loses all these.

spgarbet commented:
I think this will also be required:

```
for(i in names(Records_wide))
{
  if(i %in% c(idvar, "redcap_event_name", "redcap_repeat_instrument")) next;
  attributes(Records_wide[i]) <-  attributes(Records[
    if(grepl("redcap_repeat_instance", i)) "redcap_repeat_instance" else i
  ])
}
```

spgarbet commented:
@nutterb The above snippet is not without danger, but I think it works since it should preserve the data type. Thoughts?

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+  checkmate::assert_class(x = rcon,
+                          classes = "redcapApiConnection",
+                          add = coll)
+  checkmate::assert_character(x = idvar,
```
Remove this call. Since you're using an internal function to get idvar, we have full control over its value. It's also possible for idvar to have length 2, which is appropriate if there's a secondary unique identifier.

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+  checkmate::assert_character(x = idvar,
+                              len = 1,
+                              add = coll)
+  checkmate::assert_subset(x = idvar, names(Records))
```
Likewise, this isn't necessary.

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+  # Check parameters passed to function
+  coll <- checkmate::makeAssertCollection()
+  
+  checkmate::assert_data_frame(x = Records)
```
Please include the `add = coll` argument.

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+                              add = coll)
+  checkmate::assert_subset(x = idvar, names(Records))
+  checkmate::reportAssertions(coll)
+  
```
We haven't really worked with data access groups yet, but I think that should be included among the variables that are considered ID variables--that is, I don't think we would want `redcap_data_access_group.1` and `redcap_data_access_group.2`. 

Should we consider adding:
```
  form_ids <- c(idvar, "redcap_data_access_group", 
                "redcap_event_name", "redcap_repeat_instrument")
  form_ids <- form_ids[form_ids %in% names(Records)]
```

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+      
+      Records_wide <- reshape(
+        data = Records,
+        idvar = c(idvar, "redcap_event_name", "redcap_repeat_instrument"),
```
change to `idvar = form_ids` if accepting my previous suggestion

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+  # check that redcap_repeat_instrument is present and that all values are the same
+  if ("redcap_repeat_instrument" %in% colnames(Records) && all(!is.na(Records$redcap_repeat_instrument))) {
+    if (length(unique(Records$redcap_repeat_instrument)) == 1) {
+      cat("Great! The redcap_repeat_instrument column has unique values.\n")
```
`cat` calls cannot be suppressed by the user. There's an unwritten rule in the development community to avoid them (but they are great for development and bug hunting).  I would remove this line.

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+      stop("The redcap_repeat_instrument column does not have unique values.\n")
+    }
+  } else {
+    cat("The redcap_repeat_instrument column is not present.\n")
```
Change `cat` to `message`

spgarbet replied:
Drop cat and don't print a message at all.

spgarbet replied:
or provide a quiet argument to disable.

obregos replied:
dropping the cat call

nutterb reviewed (commented):
I love this. It is both direct and straightforward. It took very little cognitive effort for me to understand what you were intending to do. 

Most of my comments are very nit picky things. Which is good. When we start nit picking, you know you've hit the target!

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,50 @@
+    if (length(unique(Records$redcap_repeat_instrument)) == 1) {
+      cat("Great! The redcap_repeat_instrument column has unique values.\n")
+      
+      Records_wide <- reshape(
```
The style throughout the rest of the package would declare this `RecordsWide` (camel case)

nutterb commented:
> for(i in names(Records_wide))
> {
>   if(i %in% c(idvar, "redcap_event_name", "redcap_repeat_instrument")) next;
>   j <- strsplit(i, "\\.")[[1]]
>   j <- paste(j[1:(length(j)-1)], collapse=".")
>   attributes(Records_wide[[i]]) <-  attributes(Records[[j]])
>   ])
> }

You're close, but you're still losing the labels on the id variables. How about

```
for(rw in names(Records_wide)){
  if(rw %in% c(idvar, "redcap_event_name", "redcap_repeat_instrument")){
    attributes(Records_wide[[rw]]) <- attributes(Records[[rw]])
  } else {
    r <- sub("^(.*)[.].*", "\\1", rw) # Remove text from last period and after.
    attributes(Records_wide[[rw]]) <-  attributes(Records[[r]])
  }
}
```

(I prefer the `rw` and `r` iterators here, but the `i` and `j` work just as well.  `i` and `j` just always make me think of integer iterators, while these are iterating over characters)

spgarbet commented:
It's a tricky one. If that loop works. I'm good with it.

obregos commented:
Removed cat calls, and unnecessary `checkmate` calls. Added camel casing for `RecordsWide`. 
Added the `form_ids` declaration and updated instances of `idvar` to call `form_ids` instead.
Preserved column label attribute with the loop provided by @nutterb.

nutterb commented on R/widerRepeated.R:
```diff
@@ -0,0 +1,63 @@
+  checkmate::assert_data_frame(x = Records, 
+                               add = coll)
+  checkmate::assert_class(x = rcon,
+                          classes = "redcapApiConnection",
```
`classes = "redcapConnection"`

This will allow the function to work with offline connections as well.

nutterb commented:
We still need this at about line 44

```
if(!is.null(attr(Records, "invalid")))
{
  attr(RecordsWide, "invalid") <- attr(Records, "invalid")
  attr(RecordsWide, "invalid")$row <- NA # Since we can't tell where these landed
}
```

But that should be the end of it.

nutterb commented:
Oops.  one more thing

I think you should add yourself as a contributor in the DESCRIPTION file. This pull request can't possibly be complete without that. :)

spgarbet reviewed (approved):
All requests are addressed, tests are passing, and CHECK passes.
